File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
import sys
from collections import deque
input=sys.stdin.readline



while(True):
    str = input().rstrip()
    stack = []
    small = 0
    big = 0
    
    # 종료 조건
    if(str == '.'):
        break

    for c in str:
        if(c == '(' or c == ')' or c == '[' or c == ']'):
            stack.append(c)
    
    
    isDeleted = True
    while(isDeleted):
        isDeleted = False

        for i in range(len(stack)-1):
            if stack[i] =='(' and stack[i+1] == ')':
                stack.pop(i+1)
                stack.pop(i)
                isDeleted = True
                break
            elif stack[i] =='[' and stack[i+1] == ']':
                stack.pop(i+1)
                stack.pop(i)
                isDeleted = True
                break

    if stack:
        print("no")
    else:
        print("yes")   
        
#([(([([])()(())]))])

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.cc
#include <bits/stdc++.h>

using namespace std;

int main()
{
	ios::sync_with_stdio(0);
	cin.tie(0);
	string input;
	while (true)
	{
		getline(cin, input);
		if (input[0] == '.')
			return 0;

		stack<char> st;
		for (char c : input)
		{
			if (c == '(' || c=='[')
				st.push(c);
			else if (c == ')')
			{
				if (st.empty())
				{
					cout << "no" << '\n';
					break;
				}
				else
				{
					char cc = st.top();
					if (cc == '(')
						st.pop();
					else
					{
						cout << "no" << '\n';
						break;
					}
				}
			}
			else if (c == ']')
			{
				if (st.empty())
				{
					cout << "no" << '\n';
					break;
				}
				else
				{
					char cc = st.top();
					if (cc == '[')
						st.pop();
					else
					{
						cout << "no" << '\n';
						break;
					}
				}
			}
			else if (c == '.')
			{
				if (st.empty())
					cout << "yes" << '\n';
				else
					cout << "no" << '\n';
			}
		}
	}

	return 0;
}